### ISSUE TITLE
Could the loss code be more clear?

### ISSUE THREAD

**Issue by JunMa11** [LIVIAETS/miccai_weakly_supervised_tutorial#2]
Dear @HKervadec ,

Thank for updating the loss code. However, compared to the old version, the new loss code is hard to follow.

https://github.com/LIVIAETS/miccai_weakly_supervised_tutorial/blob/de1a52ddfacfb35ee9934cd02b42b7d0d5c29744/TutorialCode_WeakSup/losses.py#L27-L30

https://github.com/LIVIAETS/miccai_weakly_supervised_tutorial/blob/de1a52ddfacfb35ee9934cd02b42b7d0d5c29744/TutorialCode_WeakSup/losses.py#L75-L79

Moreover, why do we need to define the `backward()` manually? In the old version, the gradients of loss can be computed automatically by pytorch.

The old loss code is very clear.
```python
# weak cross entropy
    def __call__(self, probs, weak_target):
        assert simplex(probs)
        assert sset(weak_target, [0, 1])

        log_p = (probs[:, self.idc, ...] + 1e-10).log()
        mask = weak_target[:, None, ...].float()

        loss = - einsum(f"bkwh,bkwh->", mask, log_p)
        loss /= mask.sum() + 1e-10

        return loss

# size loss
    def __call__(self, input):
        assert simplex(input)

        pred_size = einsum("bcwh->bc", input)[:, 1]
        target_size = 7845

        loss = (pred_size - target_size) ** 2

        return loss.mean() / 1000000

```

**Would it be possible for you to make the loss code more clear? (like the old version)**

Looking forward to your reply.

Kindest regards,
Jun

**Comment by HKervadec** [comment]
I will fix this and #1 in the next few days. Some results are different because the demo used Pytorch 0.4 (quite old), whereas the initially submitted one is for the last version of Pytorch. Some minor difference in the framework might have been introduced, explaining the different results ; I will have to hunt that.

**Comment by JunMa11** [comment]
Dear @HKervadec @josedolz 

is there any update here?

**Comment by HKervadec** [comment]
After some deliberation between us, we will do two folders:
* One with the PyTorch version ; which is a simplified version of the current papers code
* One with the code Jose used during the talk. It is actually the first implementation that we used in the past, which was not refactored yet and is compatible only with pytorch 0.4 

There is many people online and around me that report they cannot get exactly the same results with the lastest versions of PyTorch, so I believe it has a role in that. This is compounded by the fact this toy example is really lightweight, and thus much more sensitive such changes.

**Comment by JunMa11** [comment]
Hi @HKervadec ,

Thanks for your reply. It's ok to make two folders for different PyTorch version (0.4 and 1.0+).
My question is:

- Would it be possible for you to make the loss code more clear in PyTorch 0.4 version? 
Just like the 1.0+ version
```
# size loss
    def __call__(self, input):
        assert simplex(input)

        pred_size = einsum("bcwh->bc", input)[:, 1]
        target_size = 7845

        loss = (pred_size - target_size) ** 2

        return loss.mean() / 1000000
```

- When the final version of the two folders will be released?

**Comment by JunMa11** [comment]
> This is compounded by the fact this toy example is really lightweight, and thus much more sensitive such changes.

Could you add more layers/structures to make the code more robust?

**Comment by HKervadec** [comment]
> > This is compounded by the fact this toy example is really lightweight, and thus much more sensitive such changes.
> 
> Could you add more layers/structures to make the code more robust?

Yes, but then we would lose the goal of this toy example: to quickly show to newcomers how the whole thing works ; while running in 10 minutes on a laptop. It required a few trade-offs to fit those constraints. 

Adding more layers, or more data samples, would be contrary to that goal. Still, users are free to tweak the details (including the CNN architecture) and see the effects on it. All we are providing here is a toy skeleton that can be used as a basis for another project.

More complete implementations (with also real CNN architectures) are available on other repositories:
* [Constrained CNN MedIA paper](https://github.com/LIVIAETS/constrained_cnn)
* [Extended log-barrier paper](https://github.com/LIVIAETS/semi_curriculum)
* [Semi-curriculum learning](https://github.com/LIVIAETS/extended_logbarrier)

> Would it be possible for you to make the loss code more clear in PyTorch 0.4 version?

This is really old code, that we do not use anymore ; and we do not intent to update it anymore. If you dig through it (and redo the derivatives by hand), you would realize that it boils down to the current PyTorch implementation. 

Currently we will leave it for historical reasons, but you should focus on the new implementation ; which is the one I will update in the future.
